File: Segundo/PDOO/Tema 1/ProyectoCivitasRuby/Civitas/TituloPropiedad.rb
# encoding:utf-8

require_relative "Jugador"

module Civitas

    class TituloPropiedad

        attr_reader :nombre
        attr_reader :hipotecado
        attr_reader :nCasas
        attr_reader :nHoteles
        attr_reader :precioCompra
        attr_reader :precioEdificar
        attr_reader :propietario
        attr_reader :precioBaseHipoteca

        def initialize(nombre, precioBaseAlquiler, factorRevalorizacion, precioBaseHipoteca, precioCompra, precioEdificar )
            @nombre = nombre
            @precioBaseAlquiler = precioBaseAlquiler
            @factorRevalorizacion = factorRevalorizacion
            @precioBaseHipoteca = precioBaseHipoteca
            @precioCompra = precioCompra
            @precioEdificar = precioEdificar
            @propietario = nil
            @nCasas = 0
            @nHoteles = 0
            @hipotecado = false
            @FactorInteresHipoteca = 1.1
        end

        def to_String
            devolver = "La calle "+ @nombre + " tiene un precio base de alquiler de " + @precioBaseAlquiler.to_s + ",\nun factor de revalorización de " + @factorRevalorizacion.to_s
            devolver += ", un precio base de hipotecado de " + @precioBaseHipoteca.to_s + ",\nun precio de compra de " + @precioCompra.to_s
            devolver += " y un precio por edificar de " + @precioEdificar.to_s + ".\n\nActualmente tiene " + @nCasas.to_s + " casas y " + @nHoteles.to_s + " hoteles y se encuentra "
            if (@hipotecado)
                devolver += "hipotecada."
            else
                devolver += "no hipotecada."
            end

            return devolver
        end

        def getPrecioAlquiler
            if (@propietario.encarcelado || @hipotecado)
                return 0
            else
                return (@PrecioBaseAlquiler*(1+(@nCasas*0.5) + (@nHoteles*2.5)))
            end
        end

        def getImporteCancelarHipoteca
            return @precioBaseHipoteca*@FactorInteresHipoteca
        end

        def tramitarAlquiler(jugador)
            if (@propietario != jugador)
                precioAlq = getPrecioAlquiler()
                jugador.pagaAlquiler(precioAlq)
                @propietario.recibe(precioAlq)
            end
        end

        def propietarioEncarcelado
            return @propietario.encarcelado
        end

        def cantidadCasasHoteles
            return (@nCasas + @nHoteles)
        end

        def getImporteHipoteca
            return (@PrecioBaseHipoteca*(1+(@nCasas*0.5) + (@nHoteles*2.5)))
        end

        def getPrecioVenta
            return (@precioCompra + @precioEdificar*self.cantidadCasasHoteles()*@factorRevalorizacion)
        end

        def derruirCasas(n, jugador)
            if (jugador == @propietario && @nCasas >= n)
                nCasas -= n
                return true
            else
                return false
            end
        end

        def vender(jugador)
            if (jugador == @propietario && !@hipotecado)
                @propietario.recibe(self.getPrecioVenta())
                @propietario = nil
                @nCasas = 0
                @nHoteles = 0
                return true
            else
                return false
            end
        end

        def tienePropietario
            if(@propietario != nil)
                return true
            else
                return false
            end
        end

        def actualizarPropietario(jugador)
            @propietario = jugador
        end

        def esElPropietario(jugador)
            return (@propietario == jugador)
        end

        def cancelarHipoteca(jugador)
            result = false
            if @hipotecado
                if esElPropietario(jugador)
                    if jugador.paga(getImporteCancelarHipoteca())
                        @hipotecado = false
                        result = true
                    end
                end
            end
            return result
        end

        def hipotecar(jugador)
            result = false
            if (!@hipotecado && esElPropietario(jugador))
                jugador.recibe(getImporteHipoteca())
                @hipotecado = true
                result = true
            end
            return result
        end

        def comprar(jugador)
            result = false
            if !tienePropietario()
                @propietario = jugador
                result = true
                jugador.paga(@precioCompra)
            end
            return result
        end

        def construirHotel(jugador)
            result = false
            if esElPropietario(jugador)
                jugador.paga(@precioEdificar)
                @nHoteles++
                result = true
            end
            return result
        end

        def construirCasa(jugador)
            result = false
            if esElPropietario(jugador)
                jugador.paga(@precioEdificar)
                @nCasas++
                result = true
            end
            return result
        end
        
    end

end

File: Segundo/PDOO/Tema 1/Civitas/src/main/java/civitas/TituloPropiedad.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author jmramirez
 */
public class TituloPropiedad {
    
    private String nombre;
    private float PrecioBaseAlquiler;
    private double FactorRevalorizacion;
    private float PrecioBaseHipoteca;
    private float PrecioCompra;
    private float PrecioEdificar;
    private int nCasas;
    private int nHoteles;
    private Boolean Hipotecado;
    private final double FactorInteresesHipoteca = 1.1;
    private Jugador Propietario;
    private Boolean Comprada;
    
    
    TituloPropiedad(String n, float PBA, double FR, float PBH, float PC, float PE) {
        nombre = n;
        PrecioBaseAlquiler = PBA;
        FactorRevalorizacion = FR;
        PrecioBaseHipoteca = PBH;
        PrecioCompra = PC;
        PrecioEdificar = PE;
        Hipotecado = false;
        Comprada = false;
        nCasas = 0;
        nHoteles = 0;
    }
    
    public String toString() {
        String Devolver;
        Devolver = "La calle " + nombre + " tiene un precio base de alquiler " + PrecioBaseAlquiler;
        Devolver += ",\nun factor de revalorización de " + FactorRevalorizacion + ", un precio base de hipotecado " + PrecioBaseHipoteca + ",\nun precio de compra de " + PrecioCompra;
        Devolver += " y un precio por edificar de " + PrecioEdificar + ".\nActualmente tiene " + nCasas + " casas y " + nHoteles + " hoteles y se encuentra ";
        if (Hipotecado) {
            Devolver += "hipotecada.\n";
        } else {
            Devolver += "no hipotecada.\n";
        }
        
        return Devolver;
    }
    
    float getPrecioAlquiler() {
        if (Propietario.isEncarcelado() || Hipotecado) {
            return 0;
        } else {
            return (float) (PrecioBaseAlquiler*(1+(nCasas*0.5) + (nHoteles*2.5)));
        }
    }
    
    float getImporteCancelarHipoteca() {
        return (float) (PrecioBaseHipoteca*FactorInteresesHipoteca);
    }
    
    void tramitarAlquiler(Jugador jugador) {
        if (Comprada && !esElPropietario(jugador)) {
            float precioAlq = this.getPrecioAlquiler();
            jugador.pagaAlquiler(precioAlq);
            Propietario.recibe(precioAlq);
        }
    }
    
    Boolean propietarioEncarcelado() {
        return (Comprada && Propietario.isEncarcelado());
    }
    
    int cantidadCasasHoteles() {
        return (nCasas + nHoteles);
    }
    
    float getPrecioVenta() {
        return (float) (PrecioCompra + PrecioEdificar*this.cantidadCasasHoteles()*FactorRevalorizacion);
    }
    
    Boolean derruirCasas(int n, Jugador jugador) {
        if (jugador == Propietario && nCasas >= n) {
            nCasas -= n;
            return true;
        } else return false;
    }
    
    Boolean vender(Jugador jugador) {
        if (jugador == Propietario && Comprada && !Hipotecado) {
            Propietario.recibe(this.getPrecioVenta());
            Comprada = false;
            Propietario = null;
            nCasas = 0;
            nHoteles = 0;
            return true;
        } else return false;
    }
    
    void actualizarPropietario(Jugador jugador) {
        Propietario = jugador;
        Comprada = true;
    }
    
    Boolean esElPropietario(Jugador jugador) {
        return (jugador == Propietario && Comprada);
    }
    
    Boolean getHipotecado() {
        return Hipotecado;
    }
    
    public String getNombre() {
        return nombre;
    }
    
    int getNCasas() {
        return nCasas;
    }
    
    int getNHoteles() {
        return nHoteles;
    }
    
    float getPrecioCompra() {
        return PrecioCompra;
    }
    
    float getPrecioEdificar() {
        return PrecioEdificar;
    }
    
    Jugador getPropietario() {
        return Comprada ? Propietario : null;
    }
    
    Boolean tienePropietario() {
        return Comprada;
    }
    
    float getImporteHipoteca() {
        return PrecioBaseHipoteca;
    }
    
    Boolean cancelarHipoteca(Jugador jugador) {
        if (esElPropietario(jugador) && Hipotecado) {
            jugador.paga(getImporteCancelarHipoteca());
            Hipotecado = false;
            return true;
        } else
            return false;
    }
    
    Boolean hipotecar(Jugador jugador) {
        if (!Hipotecado && esElPropietario(jugador)) {
            jugador.recibe(PrecioBaseHipoteca);
            Hipotecado = true;
            return true;
        }
        return false;
    }
    
    Boolean comprar(Jugador jugador) {
        if (!tienePropietario()) {
            actualizarPropietario(jugador);
            jugador.paga(PrecioCompra);
            return true;
        }
        return false;
    }
    
    Boolean construirCasa(Jugador jugador) {
        if (esElPropietario(jugador)) {
            jugador.paga(PrecioEdificar);
            nCasas++;
            return true;
        }
        return false;
    }
    
    Boolean construirHotel(Jugador jugador) {
        if (esElPropietario(jugador)) {
            jugador.paga(PrecioEdificar);
            nHoteles++;
            return true;
        }
        return false;
    }
}
